### ISSUE TITLE
Use K8s also for local development

### ISSUE THREAD

**Issue by sschuberth** [eclipse-apoapsis/ort-server#1631]
Maintaining the Docker Compose setup for local development only is cumbersome. We should try to share as much infrastructure setup as possible with the production K8s deployment. As such, probably using any of https://github.com/k3d-io/k3d, https://github.com/kubernetes/minikube or https://github.com/kubernetes-sigs/kind should replace the current Docker Compose setup.

**Comment by mnonnenmacher** [comment]
> Maintaining the Docker Compose setup for local development only is cumbersome.

Why? Docker Compose is a lot simpler than Kubernetes. Just compare `docker-compose.yml` with the content of `scripts/helm`, and the latter is not even up-to-date and would have to be extended to work.

> We should try to share as much infrastructure setup as possible with the production K8s deployment. As such, probably using any of https://github.com/k3d-io/k3d, https://github.com/kubernetes/minikube or https://github.com/kubernetes-sigs/kind should replace the current Docker Compose setup.

I think we should provide an up-to-date Helm chart to make it easy to deploy the server to any Kubernetes cluster including a local one. However, to me there are several good arguments to keep the Docker Compose file:
* Like I said before, it is a lot simpler than the Kubernetes configuration.
* It reduces the hurdle to test the server for potential users. Many more people have Docker installed than a Kubernetes cluster at hand (even a local one), and just cloning the repository an running `docker compose up` is easy.
* It consumes less resources which is good when using it for testing during development.
* A lot of development tasks do not require the added complexity of Kubernetes and can be tested just fine in the Docker Compose environment.

**Comment by sschuberth** [comment]
> Why?

Well, it's maintaining two ways instead of one to run / deploy ORT server. Ideally, there should not be much difference between deploying locally / for development and deploying for production, besides the "target" and the respective infrastructure configurations. But both could be done with basically the same tooling.

> Docker Compose is a lot simpler than Kubernetes.

Sure. That's exactly the topic that the above-mentioned "Mini-K8s-Distributions" try to address.

> Just compare `docker-compose.yml` with the content of `scripts/helm`

Still, it would be nice to only have to maintain either, not both.

> the latter is not even up-to-date and would have to be extended to work.

True, but I thought that was planned for quite some time anyway.

> I think we should provide an up-to-date Helm chart to make it easy to deploy the server to any Kubernetes cluster including a local one.

Exactly. Maybe I'm missing something due to my limited K8s knowledge, but again I though that's exactly the purpose of the mentioned tools.

> there are several good arguments to keep the Docker Compose file:
> [...]
> Like I said before, it is a lot simpler than the Kubernetes configuration.

Maybe tutorials like [this](https://devops.datenkollektiv.de/k3d-first-contact.html) and tools like [this](https://github.com/kubernetes/kompose) help to convince you.

> Many more people have Docker installed than a Kubernetes cluster at hand

Note that e.g. k3d actually runs k3s in Docker.

> It consumes less resources which is good when using it for testing during development.

You might want to have a look at [this](https://oilbeater.com/en/2024/02/22/minikube-vs-kind-vs-k3d/#Results) table, even if there is not direct comparison to Docker Compose.

**Comment by mnonnenmacher** [comment]
> Well, it's maintaining two ways instead of one to run / deploy ORT server.

Yes, but I don't see the the Compose file causes much maintenance effort.

>> Docker Compose is a lot simpler than Kubernetes.
>
> Sure. That's exactly the topic that the above-mentioned "Mini-K8s-Distributions" try to address.

They make setting up Kubernetes easier, but not making deployments. To me deploying a Helm chart is more complex than running `docker compose up`.

>> the latter is not even up-to-date and would have to be extended to work.
>
> True, but I thought that was planned for quite some time anyway.

Yes, but making a generic Helm chart is a lot more difficult than making one for a specific environment because it needs to offer many configuration options to be usable. For example, for local development it would have to deploy a message service like RabbitMQ, in production likely some managed service would be used instead.

> Maybe tutorials like [this](https://devops.datenkollektiv.de/k3d-first-contact.html) and tools like [this](https://github.com/kubernetes/kompose) help to convince you.

Not really. I'm not against anyone updating the Helm chart and using Kubernetes for local development. But I don't like the idea of making Kubernetes a hard requirement for running the server, mainly for these reasons:

* Like mentioned earlier, I believe that requiring a Kubernetes environment to be able to test the server is bad for adoption.
* I would really like if users that do not need to handle a high load could simply set up the server on a single host as a monolith, avoiding all the hassle that Kubernetes brings.

**Comment by sschuberth** [comment]
> I would really like if users that do not need to handle a high load could simply set up the server on a single host as a monolith, avoiding all the hassle that Kubernetes brings.

Ok, that makes sense.

**Comment by sschuberth** [comment]
> I think we should provide an up-to-date Helm chart to make it easy to deploy the server to any Kubernetes cluster including a local one.

As somewhat related news to that, [JKube](https://eclipse.dev/jkube/) (which we currently do not use) [can now "run tests on your Helm charts"](https://blog.marcnuri.com/eclipse-jkube-1-18#helm).

**Comment by sschuberth** [comment]
Now that I know a little bit more about the topic, I guess this issues (or at least the first step for it) should be to make https://github.com/eclipse-apoapsis/ort-server/tree/main/scripts/helm/ort-server work (again) for MiniKube.

@sparsick, isn't getting MiniKube to run anyway something you wanted to look into?

**Comment by sparsick** [comment]
@sschuberth I started to look into

My first goal is to make it run on minikube.

**Comment by sschuberth** [comment]
> My first goal is to make it run on minikube.

Great, thanks! Afterwards, let's discuss what other goals there chould be, before you start working on them.

**Comment by sschuberth** [comment]
> I could make the ORT server run on minikube (see also the [PR](https://github.com/eclipse-apoapsis/ort-server/pull/2386)).

Ah, thanks, I had trouble finding that merged PR.

@mnonnenmacher, should we then close this, as I understand that maintaining the Helm chart for local deployment is less favorable than using the Compose environment in your view?
